File: m1-spring-boot/src/test/java/com/example/microservice1/MicroserviceApplicationTests.java
package com.example.microservice1;

import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.ArgumentCaptor;
import org.springframework.amqp.rabbit.core.RabbitTemplate;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.test.mock.mockito.MockBean;
import org.springframework.test.context.junit.jupiter.SpringExtension;
import org.springframework.test.context.ActiveProfiles;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.mockito.Mockito.*;

@ExtendWith(SpringExtension.class)
@SpringBootTest
@AutoConfigureMockMvc
@ActiveProfiles("test")
class MicroserviceApplicationTests {

    @Autowired
    private MicroserviceApplication microserviceApplication;

    @MockBean
    private RabbitTemplate rabbitTemplate;

    @Test
    void testSendMessage() {
        String queueName = "kotlinMessageQueue";
        String message = "ping";

        microserviceApplication.sendMessage(queueName, message);

        ArgumentCaptor<String> exchangeCaptor = ArgumentCaptor.forClass(String.class);
        ArgumentCaptor<String> queueCaptor = ArgumentCaptor.forClass(String.class);
        ArgumentCaptor<String> messageCaptor = ArgumentCaptor.forClass(String.class);
        verify(rabbitTemplate, times(1)).convertAndSend(exchangeCaptor.capture(), queueCaptor.capture(), messageCaptor.capture());

        assertEquals("exchange", exchangeCaptor.getValue());
        assertEquals(queueName, queueCaptor.getValue());
        assertEquals(message, messageCaptor.getValue());
    }

    @Test
    void testReceiveMessage() {
        String queueName = "springMessageQueue";
        String expectedMessage = "pong";
        when(rabbitTemplate.receiveAndConvert(queueName)).thenReturn(expectedMessage);

        String actualMessage = microserviceApplication.receiveMessage(queueName);

        assertEquals(expectedMessage, actualMessage);
        verify(rabbitTemplate, times(1)).receiveAndConvert(queueName);
    }
}

File: m2-kotlin/src/test/kotlin/com/example/microservice2/MicroserviceApplicationTests.kt
package com.example.microservice2

import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.ArgumentCaptor
import org.mockito.Mockito.*
import org.springframework.amqp.rabbit.core.RabbitTemplate
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.boot.test.mock.mockito.MockBean
import org.springframework.test.context.ActiveProfiles
import org.springframework.test.context.junit.jupiter.SpringExtension
import kotlin.test.assertEquals
import kotlinx.coroutines.runBlocking

@ExtendWith(SpringExtension::class)
@SpringBootTest
@ActiveProfiles("test")
class MicroserviceApplicationTests {

    @MockBean
    private lateinit var rabbitTemplate: RabbitTemplate

    private lateinit var microserviceApplication: MicroserviceApplication

    @Test
    fun `test send message`() = runBlocking {
        val queueName = "kotlinMessageQueue"
        val message = "ping"

        microserviceApplication.sendMessage(queueName, message)

        val exchangeCaptor = ArgumentCaptor.forClass(String::class.java)
        val queueCaptor = ArgumentCaptor.forClass(String::class.java)
        val messageCaptor = ArgumentCaptor.forClass(String::class.java)
        verify(rabbitTemplate, times(1)).convertAndSend(exchangeCaptor.capture(), queueCaptor.capture(), messageCaptor.capture())

        assertEquals("exchange", exchangeCaptor.value)
        assertEquals(queueName, queueCaptor.value)
        assertEquals(message, messageCaptor.value)
    }

    @Test
    fun `test receive message`() = runBlocking {
        val queueName = "springMessageQueue"
        val expectedMessage = "pong"
        `when`(rabbitTemplate.receiveAndConvert(queueName)).thenReturn(expectedMessage)

        val actualMessage = microserviceApplication.receiveMessage(queueName)

        assertEquals(expectedMessage, actualMessage)
        verify(rabbitTemplate, times(1)).receiveAndConvert(queueName)
    }
}
